### PR TITLE
 Added method to create backup and restore

### PR DESCRIPTION
**Pull request by snigdha-px**
**What this PR does / why we need it**:
method to create backup and restore
Validate the backup and restore
**Which issue(s) this PR fixes** (optional)
Closes #
PA-345
PA-346
PA-378
**Special notes for your reviewer**:

### CODE REVIEW

File: tests/backup/backup_test.go [hunk 1]
```diff
@@ -168,14 +186,35 @@ var _ = Describe("{BasicBackupCreation}", func() {
...
+			dash.VerifyFatal(cluster_status, api.ClusterInfo_StatusInfo_Online, "Verifying backup cluster")
+		})
+
+		Step("Taking backup of applications", func() {
+			for _, namespace := range bkpNamespaces {
+				backupName := fmt.Sprintf("%s-%s", BackupNamePrefix, namespace)
+				CreateBackup(backupName, sourceClusterName, backup_location_name, BackupLocationUID, []string{namespace},
+					labelSelectores, orgID, cluster_uid, "backup-pre-rule", pre_rule_uid, "backup-post-rule", post_rule_uid)
+				//If backup is taken without pre/post rule
+				//CreateBackup(backupName, sourceClusterName, backup_location_name, BackupLocationUID, []string{namespace},
```
Comment (kshithijiyer-px) [inline tests/backup/backup_test.go:206]: Do we really need this code?

**Review by kshithijiyer-px — DISMISSED** [review]
Just one question

File: tests/backup/backup_test.go [hunk 2]
```diff
@@ -168,14 +186,35 @@ var _ = Describe("{BasicBackupCreation}", func() {
...
 			dash.VerifyFatal(monthly_policy_status, nil, "Creating monthly schedule policy")
 		})
+
 		Step("Register cluster for backup", func() {
-			CloudCredUID = uuid.New()
-			CreateCloudCredential("azure", "azureaccount", CloudCredUID, orgID)
-			// To create cloud clustre
-			RegisterBackupCluster(orgID, "azure", CloudCredUID)
+			// To create cloud cluster
+			//CloudCredUID = uuid.New()
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:192]: remove the commented code if not required

**Review by lsrinivas-pure — CHANGES_REQUESTED** [review]

File: tests/backup/backup_test.go (hunk 2 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go]: Done

File: tests/backup/backup_test.go (hunk 1 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go]: done

File: tests/backup/backup_test.go [hunk 3]
```diff
@@ -175,19 +191,20 @@ var _ = Describe("{BasicBackupCreation}", func() {
 		})
 
 		Step("Creating rules for backup", func() {
-			log.InfoD("Creating pre rule for deployed apps")
-			pre_rule_status := CreateRuleForBackup("backup-pre-rule", "default", app_list, "pre", ps)
+			dash.Info("Creating pre rule for deployed apps")
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:194]: change it to log.InfoD ,which will push to both dash board and log file
please refer https://portworx.atlassian.net/wiki/spaces/INFRA/pages/2591817827/Framework+Integration+with+Aetos

File: tests/backup/backup_test.go [hunk 4]
```diff
@@ -1937,35 +1968,62 @@ func getProviderClusterConfigPath(provider string, kubeconfigs []string) (string
...
+			Uid:  pre_rule_uid,
+		},
+		PostExecRuleRef: &api.ObjectRef{
+			Name: post_rule_name,
+			Uid:  post_rule_uid,
+		},
+	}
+	//ctx, err := backup.GetPxCentralAdminCtx()
+	ctx, err := backup.GetAdminCtxFromSecret()
+	dash.VerifyFatal(err, nil, "Fetching px-central-admin ctx")
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:2004]: if this need to be posted to dashboard, plz use log.FailOnError

File: tests/backup/backup_test.go [hunk 5]
```diff
@@ -2028,30 +2086,44 @@ func GetNodesForBackup(backupName string, bkpNamespace string,
...
-			Cluster:          clusterName,
-			NamespaceMapping: namespaceMapping,
+	var bkp *api.BackupObject
+	var bkp_uid string
+	backupDriver := Inst().Backup
+	log.Infof("Getting the UID of the backup needed to be restored")
+	bkpEnumerateReq := &api.BackupEnumerateRequest{
+		OrgId: orgID}
+	ctx, err := backup.GetAdminCtxFromSecret()
+	dash.VerifyFatal(err, nil, "Fetching px-central-admin ctx")
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:2096]: same here

**Review by lsrinivas-pure — CHANGES_REQUESTED** [review]

File: tests/backup/backup_test.go (hunk 3 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go:194]: done

File: tests/backup/backup_test.go (hunk 4 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go]: done

File: tests/backup/backup_test.go [hunk 6]
```diff
@@ -2028,30 +2086,44 @@ func GetNodesForBackup(backupName string, bkpNamespace string,
...
+	}
+	_, err = backupDriver.CreateRestore(ctx, createRestoreReq)
+	dash.VerifyFatal(err, nil, "Creating restore")
+	restoreInspectRequest := &api.RestoreInspectRequest{
+		Name:  restoreName,
+		OrgId: orgID,
+	}
+	time.Sleep(time.Minute * 3)
+	resp, err := Inst().Backup.InspectRestore(ctx, restoreInspectRequest)
+	dash.VerifyFatal(resp.GetRestore().GetStatus().Status, api.RestoreInfo_StatusInfo_PartialSuccess, "Verifying restore")
```
[the remark this answers is not in the document]
Comment (snigdha-px) [reply tests/backup/backup_test.go]: done

File: tests/backup/backup_test.go [hunk 7]
```diff
@@ -1937,35 +1968,63 @@ func getProviderClusterConfigPath(provider string, kubeconfigs []string) (string
...
+	}
+	backupInspectRequest := &api.BackupInspectRequest{
+		Name:  backupName,
+		Uid:   bkp_uid,
+		OrgId: orgID,
+	}
+	time.Sleep(time.Minute * 2)
+	resp, err := backupDriver.InspectBackup(ctx, backupInspectRequest)
+	dash.VerifyFatal(resp.GetBackup().GetStatus().Status, api.BackupInfo_StatusInfo_Success, "Inspecting the backup taken")
+	log.FailOnError(err, "Inspecting the backup taken")
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:2026]: check for the error first then verification, in this case if there is a err resp. wil give nil pointer panic

File: tests/backup/backup_test.go [hunk 8]
```diff
@@ -2028,30 +2087,45 @@ func GetNodesForBackup(backupName string, bkpNamespace string,
...
+	_, err = backupDriver.CreateRestore(ctx, createRestoreReq)
+	log.FailOnError(err, "Creating restore")
+	restoreInspectRequest := &api.RestoreInspectRequest{
+		Name:  restoreName,
+		OrgId: orgID,
+	}
+	time.Sleep(time.Minute * 3)
+	resp, err := Inst().Backup.InspectRestore(ctx, restoreInspectRequest)
+	dash.VerifyFatal(resp.GetRestore().GetStatus().Status, api.RestoreInfo_StatusInfo_PartialSuccess, "Verifying restore")
+	log.FailOnError(err, "Verifying restore")
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:2127]: same here

**Review by lsrinivas-pure — CHANGES_REQUESTED** [review]

File: tests/backup/backup_test.go (hunk 8 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go:2125]: done

File: tests/backup/backup_test.go (hunk 7 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go:2025]: done

**Review by lsrinivas-pure — APPROVED** [review]
LGTM

**Pull request merged**